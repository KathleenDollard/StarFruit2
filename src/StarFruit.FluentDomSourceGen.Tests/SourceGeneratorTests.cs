using ApprovalTests;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.CodeAnalysis.CSharp;
using StarFruit2.Generate;
using System;
using System.IO;
using System.Linq;
using Xunit;
using Xunit.Abstractions;

[assembly: UseReporter(typeof(VisualStudioReporter))]

namespace StarFruit.FluentDomSourceGen.Tests
{
    public class SourceGeneratorTests
    {
        private readonly ITestOutputHelper _output;

        public SourceGeneratorTests(ITestOutputHelper output)
        {
            _output = output;
        }

        // TODO: Consider whether tests based on this const are redundant with those based on files. Perhaps remove CliRoot details

        // These using choices are deliberate.
        // * StarFruit2.Common is redundant with default, 
        // * System.Threading.Tasks may or may not be in the defualt generation,
        // * System.IO is not in the generation
        private const string classBasedSource = @"
using StarFruit2;
using StarFruit2.Common;
using System.Threading.Tasks;
using System.IO;

namespace Tests
{
    public class Program
    {
       static int Main(string[] args)
            => CommandSource.Run<CliRoot>(args);
    }
    public class CliRoot
    {
        // Usage: > cli-root <int> --ctor-param --string-property Fred --string-option Flintstone --bool-option
        private bool ctorParam;
        public CliRoot(bool ctorParam)
            => this.ctorParam = ctorParam;

        public string StringProperty { get; set; }

        public async Task<int> InvokeAsync(int intArg, string stringOption, bool boolOption)
        { return await Task.FromResult(0); }

    }
}";

        private const string multipleGenericAndInterfaceIndicators = @"
using StarFruit2;
using StarFruit2.Common;
using System.Threading.Tasks;
using System.IO;

namespace Tests
{
    public class Program
    {
       static int Main(string[] args)
       {
          var x = CommandSource.Create<CliRoot>();
          CommandSource.Run<CliRoot>(args);
          return 0;
       }
    }
    public class CliRoot : ICliRoot
    {
        private bool ctorParam;
        public CliRoot(bool ctorParam)
            => this.ctorParam = ctorParam;

        public string StringProperty { get; set; }

        public async Task<int> InvokeAsync(int intArg, string stringOption, bool boolOption)
        { return await Task.FromResult(0); }

    }
}";

        [Fact]
        public void CommandSource_generator_produces_correct_code()
        {
            var generatedCodeName = "CliRootCommandSource";
            var output = SourceGeneratorUtilities.GenerateAndTestSource(generatedCodeName, classBasedSource, null);
            Approvals.Verify(output);

        }

        [Fact]
        public void CommandSourceResult_generator_produces_correct_code()
        {
            var generatedCodeName = "CliRootCommandSourceResult";
            var output = SourceGeneratorUtilities.GenerateAndTestSource(generatedCodeName, classBasedSource, null);
            Approvals.Verify(output);
        }

        [Fact]
        public void CliRoot_creates_CommandSource_that_will_compile()
        {
            var generatedCodeName = "CliRootCommandSource";
            SourceGeneratorUtilities.GenerateAndTest(generatedCodeName, "CliRoot.cs", $"../../../../StarFruit.FluentDomSourceGen.Tests.Output");
        }

        [Fact]
        public void CliRoot_creates_CommandSourceResult_that_will_compile()
        {
            var generatedCodeName = "CliRootCommandSourceResult";
            SourceGeneratorUtilities.GenerateAndTest(generatedCodeName, "CliRoot.cs", $"../../../../StarFruit.FluentDomSourceGen.Tests.Output");
        }

        [Fact]
        public void CliFromMethod_creates_CommandSource_that_will_compile()
        {
            var generatedCodeName = "FormatCommandSource";
            var outPath = "../../../../StarFruit.FluentDomSourceGen.Tests.FromMethodOutput/";
            var generatedSource = SourceGeneratorUtilities.GenerateAndTest(generatedCodeName, Path.Combine(outPath,"Program.cs"), outPath);
            generatedSource.Length.Should().BeGreaterThan(1000);
        }

        [Fact]
        public void CliFromMethod_creates_CommandSourceResult_that_will_compile()
        {
            var generatedCodeName = "FormatCommandSourceResult";
            var outPath = "../../../../StarFruit.FluentDomSourceGen.Tests.FromMethodOutput/";
            var generatedSource = SourceGeneratorUtilities.GenerateAndTest(generatedCodeName, Path.Combine(outPath, "Program.cs"), outPath);
            generatedSource.Length.Should().BeGreaterThan(1000);
        }

        [Fact]
        public void CliRoot_creates_a_single_CommandSource_set_for_multiple_indicators_of_root()
        {
            var generatedCodeName = "CliRootCommandSource";
            var cliRootCompilation = SourceGeneratorUtilities.GetCliRootCompilation(multipleGenericAndInterfaceIndicators)
                ?? throw new InvalidOperationException();

            var outputPairs = SourceGeneratorUtilities.Generate<Generator>(cliRootCompilation, out var outputCompilation, out var generationDiagnostics);

            var matchingPairs = outputPairs.Where(x => x.compilationName.EndsWith($"{generatedCodeName}.generated.cs"));
            matchingPairs.Should().HaveCount(1);
            generationDiagnostics.Should().NotHaveErrors($"{generatedCodeName} - Generation diagnostics");
            outputCompilation.Should().NotHaveErrors($"{generatedCodeName} - Generation compilation");
        }



        // The intention of this test is for you to copy problematic input code into Temp.cs to allow generator debugging and access to output.
        // If this is resolving a bug, create a test for the simple part of the bug once identified.
        [Fact]
        public void Generate_CommandSource_for_debugging_when_needed()
        {
            SourceGeneratorUtilities.GenerateAndTest("CommandSource", @"..\..\..\Temp.cs", $"../../../../Temp");
        }

        // The intention of this test is for you to copy problematic input code into Temp.cs to allow generator debugging and access to output
        // If this is resolving a bug, create a test for the simple part of the bug once identified.
        [Fact]
        public void Generate_CommandSourceResult_for_debugging_when_needed()
        {
            SourceGeneratorUtilities.GenerateAndTest("CommandSourceResult", @"..\..\..\Temp.cs", $"../../../../Temp");
        }

        // The intention of this test is for you to copy problematic input code into Temp.cs to allow generator debugging and access to output
        // If this is resolving a bug, create a test for the simple part of the bug once identified.
        [Fact]
        public void Test_SyntaxReceiver_for_debugging_when_needed()
        {
            var receiver = SourceGeneratorUtilities.FindCandidatesWithSyntaxReceiver(@"..\..\..\Temp.cs");
            receiver.Candidates.Should().NotBeNullOrEmpty();
            receiver.Candidates.Should().HaveCount(1);
        }

    }
}
