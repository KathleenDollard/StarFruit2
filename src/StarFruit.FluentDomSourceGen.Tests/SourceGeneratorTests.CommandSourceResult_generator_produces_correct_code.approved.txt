using StarFruit2;
using System.CommandLine;
using StarFruit2.Common;
using System.CommandLine.Invocation;
using System.CommandLine.Parsing;
using System.Threading.Tasks;
using System.IO;

namespace Tests
{
   public class CliRootCommandSourceResult : CommandSourceResult<CliRoot>
   {
      public CliRootCommandSourceResult(ParseResult parseResult, CliRootCommandSource commandSource, int exitCode)
      : base(parseResult, (commandSource.ParentCommandSource as CommandSourceBase), exitCode)
      {
         StringPropertyOption_Result = CommandSourceMemberResult.Create(commandSource.StringPropertyOption, parseResult);
         ctorParamOption_Result = CommandSourceMemberResult.Create(commandSource.ctorParamOption, parseResult);
      }
      public CommandSourceMemberResult<string> StringPropertyOption_Result { get; set; }
      public CommandSourceMemberResult<bool> ctorParamOption_Result { get; set; }

      public override CliRoot CreateInstance()
      {
         var newItem = new CliRoot(ctorParamOption_Result.Value);
         newItem.StringProperty = StringPropertyOption_Result.Value;
         return newItem;
      }
   }
   public class InvokeAsyncCommandSourceResult : CliRootCommandSourceResult
   {
      public InvokeAsyncCommandSourceResult(ParseResult parseResult, InvokeAsyncCommandSource commandSource, int exitCode)
      : base(parseResult, (commandSource.ParentCommandSource as CliRootCommandSource), exitCode)
      {
         intArgArgument_Result = CommandSourceMemberResult.Create(commandSource.intArgArgument, parseResult);
         stringOptionOption_Result = CommandSourceMemberResult.Create(commandSource.stringOptionOption, parseResult);
         boolOptionOption_Result = CommandSourceMemberResult.Create(commandSource.boolOptionOption, parseResult);
      }
      public CommandSourceMemberResult<int> intArgArgument_Result { get; set; }
      public CommandSourceMemberResult<string> stringOptionOption_Result { get; set; }
      public CommandSourceMemberResult<bool> boolOptionOption_Result { get; set; }

   }
}
